File: app/src/main/java/com/ricamgar/notify/reminderslist/RemindersListPresenter.java
package com.ricamgar.notify.reminderslist;

import android.support.annotation.Nullable;

import com.ricamgar.notify.base.mvp.AbstractBasePresenter;
import com.ricamgar.notify.domain.reminder.model.Reminder;
import com.ricamgar.notify.domain.reminder.usecase.DeleteReminder;
import com.ricamgar.notify.domain.reminder.usecase.GetReminders;
import com.ricamgar.notify.domain.reminder.usecase.MarkReminderAsDone;
import com.ricamgar.notify.domain.reminder.usecase.MarkReminderAsTodo;

import java.util.List;

import javax.inject.Inject;

public class RemindersListPresenter
        extends AbstractBasePresenter<List<Reminder>> {

    private final GetReminders getRemindersUseCase;
    private final MarkReminderAsDone markReminderAsDoneUseCase;
    private final MarkReminderAsTodo markReminderAsTodo;
    private final DeleteReminder deleteReminder;
    private int position;

    @Inject
    public RemindersListPresenter(GetReminders getRemindersUseCase,
                                  MarkReminderAsDone markReminderAsDoneUseCase,
                                  MarkReminderAsTodo markReminderAsTodo,
                                  DeleteReminder deleteReminder) {
        this.getRemindersUseCase = getRemindersUseCase;
        this.markReminderAsDoneUseCase = markReminderAsDoneUseCase;
        this.markReminderAsTodo = markReminderAsTodo;
        this.deleteReminder = deleteReminder;
    }

    @Override
    public void attachToView(BaseView view, @Nullable List<Reminder> viewModel) {
        super.attachToView(view, viewModel);
        if (viewModel.isEmpty()) {
            addSubscription(getRemindersUseCase.history(position == 1)
                    .execute()
                    .subscribe(
                            reminders -> {
                                viewModel.clear();
                                viewModel.addAll(reminders);
                                ((View) view).setReminders(reminders);
                            },
                            ((View) view)::showError
                    ));
        } else {
            ((View) view).setReminders(viewModel);
        }
    }

    public void reminderMarked(Reminder reminder, boolean done) {
        if (done) {
            markReminderAsDoneUseCase.setReminder(reminder);
            markReminderAsDoneUseCase
                    .execute()
                    .subscribe();
        } else {
            markReminderAsTodo.setReminder(reminder);
            markReminderAsTodo
                    .execute()
                    .subscribe();
        }
    }

    public void reminderDeleted(Reminder reminder) {
        deleteReminder.setReminder(reminder);
        deleteReminder
                .execute()
                .subscribe();
    }

    public RemindersListPresenter setPosition(int position) {
        this.position = position;
        return this;
    }

    public interface View extends BaseView {
        void setReminders(List<Reminder> reminders);

        void showError(Throwable throwable);
    }
}

File: app/src/main/java/com/ricamgar/notify/reminderslist/RemindersListPresenter.kt
package com.ricamgar.notify.reminderslist

import com.ricamgar.notify.base.mvp.AbstractBasePresenter
import com.ricamgar.notify.domain.reminder.model.Reminder
import com.ricamgar.notify.domain.reminder.usecase.DeleteReminder
import com.ricamgar.notify.domain.reminder.usecase.GetReminders
import com.ricamgar.notify.domain.reminder.usecase.MarkReminderAsDone
import com.ricamgar.notify.domain.reminder.usecase.MarkReminderAsTodo
import io.reactivex.Scheduler

import javax.inject.Inject
import javax.inject.Named

internal class RemindersListPresenter
@Inject constructor(
  private val getRemindersUseCase: GetReminders,
  private val markReminderAsDoneUseCase: MarkReminderAsDone,
  private val markReminderAsTodo: MarkReminderAsTodo,
  private val deleteReminder: DeleteReminder,
  @Named("io") ioThread: Scheduler,
  @Named("main") mainThread: Scheduler
) : AbstractBasePresenter<MutableList<Reminder>>(ioThread, mainThread) {

  private var position: Int = 0

  override fun attachToView(view: BaseView, viewModel: MutableList<Reminder>?) {
    super.attachToView(view, viewModel)
    if (viewModel != null && viewModel.isEmpty()) {
      addDisposable(getRemindersUseCase.execute(position == 1)
        .compose(flowableSchedulers())
        .subscribe(
          { reminders ->
            viewModel.clear()
            viewModel.addAll(reminders)
            (view as View).setReminders(reminders)
          },
          { (view as View).showError(it) }
        ))
    } else {
      (view as View).setReminders(viewModel)
    }
  }

  fun reminderMarked(reminder: Reminder, done: Boolean) {
    if (done) {
      markReminderAsDoneUseCase.execute(reminder)
        .subscribe()
    } else {
      markReminderAsTodo.execute(reminder)
        .subscribe()
    }
  }

  fun reminderDeleted(reminder: Reminder) {
    deleteReminder.execute(reminder.id)
      .subscribe()
  }

  fun setPosition(position: Int): RemindersListPresenter {
    this.position = position
    return this
  }

  interface View : AbstractBasePresenter.BaseView {
    fun setReminders(reminders: List<Reminder>?)

    fun showError(throwable: Throwable)
  }
}
